content/blog/2022/2022-08-15-admin-operations-with-nodejs-launchdarkly.md

```diff
@@ -0,0 +1,699 @@
+
+In the LaunchDarkly UI, it looks like this:
+
+{{% image alt="Multivariate log-level" src="images/posts/nodejs-backend-launchdarkly/Log_Level_Variations.png" %}}
+
+Next, we can define targeting values that would deliver one of the above defined multivariate strings:
+
+{{% image alt="Multivariate targeting rule" src="images/posts/nodejs-backend-launchdarkly/Targeting_Log_Level.png" %}}
+
+Note that we're not using targeting rules that target individual users, because our log level is a global feature flag that is independent of specific users.
+
+Now that we have our multivariate feature flag defined, we update our existing logger class from above with few new methods that read the log level from the feature flag variation at runtime and updates in the console messages:
+
+```javascript
+import { format } from 'date-fns';
+import padEnd from 'lodash/padEnd.js';
+import capitalize from 'lodash/capitalize.js';
+
+const LEVELS = { debug: 10, log: 20, warn: 30, error: 40 };
+let currentLogLevel = LEVELS['debug'];
+
+class DynamicLogger {
+  constructor( module, ldClient, flagKey, user ) {
+    this.module = module ? module : '';
+
+    this.debug = this.debug.bind(this);
+    this.info = this.info.bind(this);
+    this.warn = this.warn.bind(this);
+    this.error = this.error.bind(this);
+    this.writeToConsole = this.writeToConsole.bind(this);
+    this.ldClient = ldClient;
+	this.flagKey = flagKey;
+	this.user = user;
+	this.previousLevel = null; 
+  }
+
+  static setLogLevel(level) {
+    currentLogLevel = LEVELS[level];
+  }
+
+  static get(module) {
+    return new Logger(module);
+  }
+
+  writeToConsole(level, message) {
+    if (LEVELS[level] >= currentLogLevel) {
+      const dateTime = format(new Date(), 'MM-dd-yyyy HH:mm:ss:SSS');
+      const formattedLevel = padEnd(capitalize(level), 5);
+      const formattedMessage = `${dateTime} ${formattedLevel} [${
+        this.module
+      }] ${message}`;
+      console[level](formattedMessage, '');
+    }
+  }
+
+  async debug( message ) { 
+    if ( await this._presentLog( 'debug' ) ) {
+        this.writeToConsole('debug', message); 
+    } 
+  }
+
+  async error( message ) { 
+    if ( await this._presentLog( 'error' ) ) {
+        this.writeToConsole('error', message); 
+    } 
+  }
+
+  async info( message ) { 
+    if ( await this._presentLog( 'info' ) ) {
+        this.writeToConsole('info', message); 
+    } 
+  }
+
+  async warn( message ) { 
+    if ( await this._presentLog( 'warn' ) ) {
+        this.writeToConsole('warn', message); 
+    }
+  }
+
+  async _presentLog( level ) {
+
+    const minLogLevel = await this.ldClient.variation(
+        this.flagKey,
+        {
+           key: this.user
+        },
+        'debug' // Default / fall-back value if LaunchDarkly unavailable.
+    );
+
+    if ( minLogLevel !== this.previousLevel ) { 
+       console.log( `Switching to log-level: ${ minLogLevel }` ); 
+    }
+        
+    switch ( this.previousLevel = minLogLevel ) {
```

Comment:
I don't understand the logic of this `switch` statement. Why is the `_presentLog()` function returning a boolean at all? The `writeToConsole` function already checks the log level (and much more elegantly than this switch statement):

```
if (LEVELS[level] >= currentLogLevel)
```

Can we just ditch this switch construct, make the `_presentLog()` function not return anything and remove all the `if`s from the `error()` etc. functions?

Reply:
Instead we need to remove that `if (LEVELS[level] >= currentLogLevel)` check. The `_presentLog()` basically listens for any changes in the log level dynamically and pass the present log level. This switch statement checks the minimum log level and pass that information to each `if` statements dynamically.

I have removed that `if (LEVELS[level] >= currentLogLevel)` check in `writeToConsole()` function and retaining the `_presentLog()` to listen for any changes in the value of the flag.

Reply:
If you do it this way, then please remove `LEVELS` constant. Also, we don't need the `setLogLevel()` function, because the log level is controlled by the feature flag (in the `_presentLog()` function)

Reply:
@thombergs done.. I have made relevant changes. Please review.